pkg/kubelet/kubelet.go

```diff
@@ -2845,6 +2856,84 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxContainerCheckpoints goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxContainerCheckpoints'.
+func (kl *Kubelet) handleMaxContainerCheckpoints(podUID types.UID, podFullName, containerName string, options *runtimeapi.CheckpointContainerRequest) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxContainerCheckpointsLock.Lock()
+	defer kl.maxContainerCheckpointsLock.Unlock()
+
+	checkpointDirectory, err := os.ReadDir(kl.getCheckpointsDir())
+	if err != nil {
+		return fmt.Errorf("error reading checkpoint archive directory %s: %w", kl.getCheckpointsDir(), err)
+	}
+
+	regexString := fmt.Sprintf(`checkpoint-%s-%s-\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.*tar`, podFullName, containerName)
+
+	re, err := regexp.Compile(regexString)
+	if err != nil {
+		return fmt.Errorf("error compiling regex %s: %w", regexString, err)
+	}
+
+	var checkpointArchives = []fs.FileInfo{}
+	for _, archive := range checkpointDirectory {
+		// only select checkpoint archives which have been created from the same
+		// namespace/pod/container combination
+		if re.MatchString(archive.Name()) {
+			archiveFileinfo, err := archive.Info()
+			if err != nil {
+				return fmt.Errorf("error getting file info %s:%w", archive.Name(), err)
+			}
+			checkpointArchives = append(
+				checkpointArchives,
+				archiveFileinfo,
+			)
+		}
+	}
+
+	checkpointArchivesCounter := len(checkpointArchives)
+	if checkpointArchivesCounter <= int(kl.maxContainerCheckpoints) {
+		// Nothing to clean up, return early.
+		return nil
+	}
+
+	sort.Slice(checkpointArchives, func(i, j int) bool {
+		stats := make([]time.Time, 2)
+		for ind, archive := range []fs.FileInfo{checkpointArchives[i], checkpointArchives[j]} {
+			stats[ind] = archive.ModTime()
+		}
+		return stats[0].Before(stats[1])
+	})
+
+	for _, archive := range checkpointArchives[:checkpointArchivesCounter-int(kl.maxContainerCheckpoints)] {
+		archiveWithPath := filepath.Join(
+			kl.getCheckpointsDir(),
+			archive.Name(),
+		)
+		klog.InfoS("More than maxContainerCheckpoints", "container", containerName, "numberOfCheckpoints", checkpointArchivesCounter, "maxContainerCheckpoints", kl.maxContainerCheckpoints)
```

Comment:
This message is identical for all iterations. Can you move it out of the loop?

Reply:
moved